table/tables/partition_test.go

```diff
@@ -700,6 +702,1550 @@ func TestIssue31629(t *testing.T) {
+	result.CheckNotContainMore([]string{"partition:p1", "partition:p0", "partition:p3"})
+	tk.MustQuery("SELECT * FROM tkey12_2 WHERE col2 = 2 and col3 IS NULL").Check(testkit.Rows("1 2 <nil> 2"))
+	result = tk.MustQuery("EXPLAIN SELECT * FROM tkey12_2 WHERE col2 = 2")
+	result.CheckContainMore([]string{"partition:p0", "partition:p1", "partition:p2", "partition:p3"})
+	tk.MustQuery("SELECT * FROM tkey12_2 WHERE col2 = 2").Check(testkit.Rows("1 2 <nil> 2"))
+	tk.MustQuery("EXPLAIN SELECT * FROM tkey12_2 WHERE col2 = 2").CheckContainMore([]string{"partition:p0", "partition:p1", "partition:p2", "partition:p3"})
+	tk.MustQuery("SELECT * FROM tkey12_2 WHERE col2 IS NULL")
+	tk.MustQuery("EXPLAIN SELECT * FROM tkey12_2 WHERE col2 IS NULL").CheckContainMore([]string{"partition:p0", "partition:p1", "partition:p2", "partition:p3"})
+	// Get the partition information from information_schema.partitions
+	result = tk.MustQuery("select PARTITION_NAME,PARTITION_ORDINAL_POSITION,PARTITION_METHOD,PARTITION_EXPRESSION " +
+		"FROM information_schema.partitions where TABLE_NAME = 'tkey12_2'")
+	result.Check(testkit.Rows("p0 1 KEY `col2`,`col3`", "p1 2 KEY `col2`,`col3`", "p2 3 KEY `col2`,`col3`", "p3 4 KEY `col2`,`col3`"))
+
+	// This tests caculating the boundary partition ID when it prunes partition table
+	tk.MustExec("create table tkey16 (a int) partition by key (a) partitions 12")
+	tk.MustExec("insert into tkey16 values (0), (1), (2), (3)")
+	tk.MustExec("insert into tkey16 select a + 4 from tkey16")
+	tk.MustExec("insert into tkey16 select a + 8 from tkey16")
+	tk.MustExec("select * from information_schema.partitions where partition_name is not null")
+}
+
+func TestKeyPartitionWithDifferentCharsets(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("create database partitiondb4")
+	defer tk.MustExec("drop database partitiondb4")
+	tk.MustExec("use partitiondb4")
+
+	tk.MustExec("CREATE TABLE tkey29 (" +
+		"col1 INT NOT NULL," +
+		"col2 DATE NOT NULL," +
+		"col3 VARCHAR(12) NOT NULL," +
+		"col4 INT NOT NULL," +
+		"UNIQUE KEY (col3)" +
+		") CHARSET=utf8mb4 COLLATE=utf8mb4_bin " +
+		"PARTITION BY KEY(col3) " +
+		"PARTITIONS 4")
+	// ignore tail spaces
+	err := tk.ExecToErr("INSERT INTO tkey29 VALUES(1, SYSDATE(), 'linpin', 1), (1, SYSDATE(), 'linpin ', 5)")
+	require.Regexp(t, "Duplicate entry 'linpin ' for key 'tkey29.col3'", err)
+	// case sensitive
+	tk.MustExec("INSERT INTO tkey29 VALUES(3, SYSDATE(), 'abc', 1), (4, SYSDATE(), 'ABC ', 5)")
+
+	tk.MustExec("CREATE TABLE tkey30 (" +
+		"col1 INT NOT NULL," +
+		"col2 DATE NOT NULL," +
+		"col3 VARCHAR(12) NOT NULL," +
+		"col4 INT NOT NULL," +
+		"UNIQUE KEY (col3)" +
+		") CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci " +
+		"PARTITION BY KEY(col3) " +
+		"PARTITIONS 4")
+	// case insensitive
+	err = tk.ExecToErr("INSERT INTO tkey30 VALUES(1, SYSDATE(), 'linpin', 1), (1, SYSDATE(), 'LINPIN', 5)")
```

Comment:
I just noticed that SYSDATE() will generate a warning for DATE columns. No need to change it, just an observation from my side.

Reply:
It doesn't gerneate a warning in my enviroment, as follows, but I will change to a specific date string.
`MySQL [partitiondb]> INSERT INTO tkey0 VALUES(1, SYSDATE(), 1, 1), (2, SYSDATE(), 2, 2), (3, SYSDATE(), 3, 3), (4, SYSDATE(), 4, 4);
Query OK, 4 rows affected (0.02 sec)
Records: 4  Duplicates: 0  Warnings: 0`